### PR TITLE
add TimestampIndex changes needed for bitcore block explorer

### PR DESCRIPTION
**Pull request by LarryRuane**
Addresses #3708, follow-on to #3849, no tests yet. This is the third of 3 indices that we're adding.

### CODE REVIEW

**Comment by LarryRuane** [comment]
Because this PR is layered upon three earlier ones (#3832, #3837, #3849) so includes all commits from those PRs (until they get merged). Please only review here only commits made on or after Feb 21.

File: src/Makefile.am [hunk 1]
```diff
@@ -114,6 +114,7 @@ LIBZCASH_H = \
 .PHONY: FORCE collate-libsnark check-symbols check-security
 # bitcoin core #
 BITCOIN_CORE_H = \
+  addressindex.h \
```
Comment (LarryRuane) [inline src/Makefile.am]: I only just discovered that these changes to this file are needed (apparently, although I haven't seen any ill-effects by not adding these lines). The header file names listed here are in alphabetical order.

File: src/init.cpp [hunk 2]
```diff
@@ -340,8 +340,7 @@ std::string HelpMessage(HelpMessageMode mode)
 
     string strUsage = HelpMessageGroup(_("Options:"));
     strUsage += HelpMessageOpt("-?", _("This help message"));
-    strUsage += HelpMessageOpt("-addressindex", strprintf(_("bitcore: Maintain a full address index, used to query for the balance, txids and unspent outputs for addresses (default: %u)"), false));
-    strUsage += HelpMessageOpt("-spentindex", strprintf(_("Maintain a full spent index, used to query the spending txid and input index for an outpoint (default: %u)"), false));
+    strUsage += HelpMessageOpt("-addressindex", strprintf(_("bitcore: Maintain a full address index, used to query for the balance, txids and unspent outputs for addresses (default: %u)"), 0));
```
Comment (LarryRuane) [inline src/init.cpp]: I moved these to maintain alphabetical order (should have been done in earlier commits).

File: src/main.cpp [hunk 3]
```diff
@@ -2774,6 +2775,26 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+            logicalTS = prevLogicalTS + 1;
+            LogPrintf("%s: Previous logical timestamp is newer Actual[%d] prevLogical[%d] Logical[%d]\n", __func__, pindex->nTime, prevLogicalTS, logicalTS);
+        }
+
+        if (!pblocktree->WriteTimestampIndex(CTimestampIndexKey(logicalTS, pindex->GetBlockHash())))
+            return AbortNode(state, "Failed to write timestamp index");
+
+        if (!pblocktree->WriteTimestampBlockIndex(CTimestampBlockIndexKey(pindex->GetBlockHash()), CTimestampBlockIndexValue(logicalTS)))
+            return AbortNode(state, "Failed to write blockhash index");
+    }
```
Comment (LarryRuane) [inline src/main.cpp]: source for the above set of added lines: https://github.com/bitpay/bitcoin/blob/1f0598705c3176f70325cd56a5446fec3ca4f13c/src/main.cpp#L2761

File: src/txdb.cpp [hunk 4]
```diff
@@ -388,6 +390,57 @@ bool CBlockTreeDB::UpdateSpentIndex(const CSpentIndexVec &vect) {
...
+{
+    CTimestampBlockIndexValue(lts);
+    if (!Read(std::make_pair(DB_BLOCKHASHINDEX, hash), lts))
+        return false;
+
+    ltimestamp = lts.ltimestamp;
+    return true;
+}
+
+
```
Comment (LarryRuane) [inline src/txdb.cpp]: source for the above set of lines: https://github.com/bitpay/bitcoin/blob/1f0598705c3176f70325cd56a5446fec3ca4f13c/src/txdb.cpp#L269

**Comment by bitcartel** [comment]
@LarryRuane Can you rebase this PR on top of the latest version of #3849 which has the fix?

**Comment by LarryRuane** [comment]
> @LarryRuane Can you rebase this PR on top of the latest version of #3849 which has the fix?

@bitcartel, done, thanks for the reminder. (I also rebased the next PR in the sequence, #3926.)

**Comment by LarryRuane** [comment]
@bitcartel, please don't merge this yet, we need to merge #3966 first, and I'll rebase this current PR on top of it (there will be a trivial merge conflict since this PR actually includes that change).

**Comment by bitcartel** [comment]
Once #3849 is merged into master, this PR can be rebased on master, and then merged.

**Comment by LarryRuane** [comment]
@bitcartel, @mdr0id, rebased on the latest master, which includes the spentindex changes. This PR should be ready to merge.

**Comment by bitcartel** [comment]
@zkbot r+

**Pull request merged**